### PR TITLE
[SPARK-33402][CORE] Jobs launched in same second have duplicate MapReduce JobIDs

### PR DESCRIPTION
**Pull request by steveloughran**
### What changes were proposed in this pull request?

1. Applies the SQL changes in SPARK-33230 to SparkHadoopWriter, so that `rdd.saveAsNewAPIHadoopDataset` passes in a unique job UUID in `spark.sql.sources.writeJobUUID`
1. `SparkHadoopWriterUtils.createJobTrackerID` generates a JobID by appending a random long number to the supplied timestamp to ensure the probability of a collision is near-zero.
1. With tests of uniqueness, round trips and negative jobID rejection.

### Why are the changes needed?

Without this, if more than one job is started in the same second *and the committer expects application attempt IDs to be unique* is at risk of clashing with other jobs.

With the fix,

* those committers which use the ID set in `spark.sql.sources.writeJobUUID` as a priority ID will pick that up instead and so be unique.
* committers which use the Hadoop JobID for unique paths and filenames will get the randomly generated jobID.  Assuming all clocks in a cluster in sync, the probability of two jobs launched in the same second has dropped from 1 to 1/(2^63)

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit tests

There's a new test suite SparkHadoopWriterUtilsSuite which creates jobID, verifies they are unique even for the same timestamp and that they can be marshalled to string and parsed back in the hadoop code, which contains some (brittle) assumptions about the format of job IDs.

Functional Integration Tests

1. Hadoop-trunk built with [HADOOP-17318], publishing to local maven repository
1. Spark built with hadoop.version=3.4.0-SNAPSHOT to pick up these JARs.
1. Spark + Object store integration tests at [https://github.com/hortonworks-spark/cloud-integration](https://github.com/hortonworks-spark/cloud-integration) were built against that local spark version
1. And executed against AWS london.

The tests were run with `fs.s3a.committer.require.uuid=true`, so the s3a committers fail fast if they don't get a job ID down. This showed that `rdd.saveAsNewAPIHadoopDataset` wasn't setting the UUID option. It again uses the current Date value for an app attempt -which is not guaranteed to be unique.

With the change applied to spark, the relevant tests work, therefore the committers are getting unique job IDs.

### CODE REVIEW

**Comment by steveloughran** [comment]
FYI, @dongjoon-hyun @rdblue -this is the rdd version of the sql-side patch. Tests downstream can now flag the absence/presence of this property when I set the committer to fail in job set up if it wasn't found

**Comment by steveloughran** [comment]
For full due diligence, one test suite did fail as Hive couldn't bring up an S3A connector instance in the isolated classloader. https://issues.apache.org/jira/browse/HADOOP-17372

**Comment by dongjoon-hyun** [comment]
Thanks, @steveloughran .
cc @sunchao , too

**Comment by steveloughran** [comment]
I'm thinking it may be best to not only do this but do better randomness in the creation of the timestamp of the artificial app attempt ID. Today: Date() is used, hence the conflict if >1 job is started in the same second. The staging committer is most vulnerable to this, but if someone uses FileOutputCommitter to the same destination dir *and* overwrite is enabled, the same conflict occurs

I'm making sure the S3a committers pick up this UUID everywhere (staging already does for the clusterfs, but not for local task attempt dir). What I'm not going to go near is the classic FileOutputCommitter, for the following reason: fear

I don't want to go anywhere near that committer as it is way too critical, and it contains deep assumptions that application attempt IDs are sequential; Hadoop MR uses that for recoverability on restarted job attempts. Spark doesn't worry about failed drivers, so doesn't need that sequential naming,

**Comment by dongjoon-hyun** [comment]
> Spark built with hadoop.version=3.4.0-SNAPSHOT to pick up these JARS. For reasons I don't understand, spark master wouldn't build with that hadoop version, but an internal spark 2.x branch was happy.

For the above, we were blocked at https://github.com/apache/spark/pull/30135 `[SPARK-29250][BUILD] Upgrade to Hadoop 3.3.0` due to https://github.com/apache/hadoop/pull/2411 `HADOOP-17324. Don't relocate org.bouncycastle in shaded client jars` and it's resolved yesterday.

File: core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriterUtils.scala [hunk 1]
```diff
@@ -37,14 +36,39 @@ private[spark]
...
+    var l1 = RAND.nextLong()
+    if (l1 < 0) {
+      l1 = -l1
+    }
+    var l2 = RAND.nextLong()
+    if (l2 < 0) {
+      l2 = -l2
+    }
+    // use leading zeros to ensure the id is always at least four digits long
+    f"$l1%04d$l2"
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriterUtils.scala:71]: Although this is a random string, this looks like a big change to me.

Comment (steveloughran) [reply core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriterUtils.scala:71]: I did think about doing something here retaining date and time. If you want a random number at the end, happy to oblige. What's key is: it must be the digits 0-9 only.

Comment (steveloughran) [reply core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriterUtils.scala]: updated the patch. Timestamp + single random long...reduces probability of collision by 2^63.

**Comment by steveloughran** [comment]
bq. For the above, we were blocked at #30135 [SPARK-29250][BUILD] Upgrade to Hadoop 3.3.0 due to apache/hadoop#2411 HADOOP-17324. Don't relocate org.bouncycastle in shaded client jars and it's resolved yesterday.

I am happy someone else is *also* trying to build trunk branches of multiple things together. It spreads the joy

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you @steveloughran .
Merged to master/3.0.

**Comment by steveloughran** [comment]
Thanks!

File: core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala [hunk 2]
```diff
@@ -70,6 +70,11 @@ object SparkHadoopWriter extends Logging {
     // Assert the output format/key/value class is set in JobConf.
     config.assertConf(jobContext, rdd.conf)
 
+    // propagate the description UUID into the jobs, so that committers
+    // get an ID guaranteed to be unique.
+    jobContext.getConfiguration.set("spark.sql.sources.writeJobUUID",
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala:75]: I'm wondering why is this needed. Can't commit protocol just get a unique job ID from the constructor parameters?

https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/internal/io/FileCommitProtocol.scala#L32-L34
